urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/filehandler/CGIHandler.java

```diff
@@ -0,0 +1,485 @@
+    /**
+     * Most file handlers operate on existing files that are located inside the local file 
+     * system (such as the default handler does for simple file delivery).
+     *
+     * Some file handlers operate on virtual file systems, which means that the request URI does
+     * not necessarily address an existing file but a symbol only the handler may know. 
+     *
+     * The global HTTP handler needs to know if to throw a MissingResourceException (resulting
+     * in a 404) if a requested file does not exists --- or if to ignore that fact and simply 
+     * continue.
+     *
+     * This method tells how to proceed.
+     *
+     * If your implementation returns true this handler will not be called at all; the request
+     * processing will directly stop raising an HTTP status 404.
+     *
+     * @return true if this file handler definitely requires existing files. The process(...) 
+     *              method will never be called if the requested file does not exist in that case.
+     **/
+    public boolean requiresExistingFile() {
+	
+	// The CGI script definitely has to exist
+	return true;
+    }
+
+     /**
+     * The 'process' method is very generic. It depends on the underlying implementation how the passed
+     * file should be processed.
+     *
+     * @param sessionID   The current session's ID.
+     * @param headers     The HTTP request headers.
+     * @param postData    The HTTP post data; if the method is not HTTP POST the 'postData' should be null
+     *                    (or empty).
+     * @param file        The requested file itself (inside the local file system).
+     * @param requestURI  The requested URI (relative to DOCUMENT_ROOT).
+     **/
+    public Resource process( UUID sessionID,
+			     HTTPHeaders headers,
+			     PostDataWrapper postData,
+			     File file,
+			     URI requestURI )
+	throws IOException,
+	       HeaderFormatException,
+	       DataFormatException,
+               UnsupportedFormatException {
+	    
+
+	// Fetch the system command (specified and constructed in the sub-class).
+	List<String> command =  this.buildCGISystemCommand( headers, postData, file, requestURI );
+
+	try {
+	    getLogger().log( Level.INFO,
+			     getClass().getName() + ".process(...)",
+			     "Processing. requestURI=" + requestURI + ". file=" + file.getAbsolutePath() );	
+	   
+	    ProcessBuilder pb = new ProcessBuilder( command );
+	
+	    this.buildCGIEnvironment( pb, headers, file, requestURI, sessionID );
+
+	    Map<String,String> environment = pb.environment();
+	    this.buildAdditionalCGIEnvironmentVars( headers, file, requestURI, environment );
+
+
+
+	    this.getLogger().log( Level.INFO,
+				  getClass().getName() + ".process(...)",
+				  "Creating a processable resource using the CGI file '" + file.getPath() + "'. System command: " + command.toString() );
+	
+	    ProcessableResource cgiOutput = 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *cgiOutput*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).